File: src/core/staleness.ts
import type { InboxSnapshot } from '@shared/ipc'

/** Past this age, the list on screen is old enough to be worth a fetch. */
const STALE_AFTER_MS = 60_000

/**
 * Whether opening the popup should spend a fetch.
 *
 * Never while one is already running. `Inbox.refresh` deliberately queues a
 * second pass behind the in-flight one rather than joining it, so a click
 * landing during the fetch that starts with the app cost two identical
 * requests back to back — and put the spinner back over a list that had
 * just arrived. A fetch already on its way is exactly what opening onto a
 * stale list wanted.
 */
export function shouldRefreshOnOpen(
  snapshot: Pick<InboxSnapshot, 'status' | 'lastUpdatedAt'>,
  now: string,
): boolean {
  if (snapshot.status === 'loading') return false
  if (snapshot.lastUpdatedAt === null) return true
  return Date.parse(now) - Date.parse(snapshot.lastUpdatedAt) > STALE_AFTER_MS
}

File: Sources/PulloverCore/Staleness.swift
import Foundation

/// Past this age, the list on screen is old enough to be worth a fetch.
public let staleAfter: TimeInterval = 60

/// Whether opening the popup should spend a fetch. Never while one is already
/// running: a refresh called mid-pass queues a second pass rather than joining.
public func shouldRefreshOnOpen(status: InboxSnapshot.Status, lastUpdatedAt: Date?, now: Date) -> Bool {
    if status == .loading { return false }
    guard let lastUpdatedAt else { return true }
    return now.timeIntervalSince(lastUpdatedAt) > staleAfter
}

public extension InboxSnapshot {
    func shouldRefreshOnOpen(now: Date) -> Bool {
        PulloverCore.shouldRefreshOnOpen(status: status, lastUpdatedAt: lastUpdatedAt, now: now)
    }
}
